File: LeetCode/1456-find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance/find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance.py
class Solution:
    def findTheCity(self, n: int, edges: List[List[int]], distanceThreshold: int) -> int:
        adj = defaultdict(list)

        for v, v_2, w in edges:
            adj[v].append((v_2, w))
            adj[v_2].append((v, w))
                
        def total_count(curr, distance_left, visited):
            for adj_elem in adj[curr]:
                adj_vertex = adj_elem[0]
                weight = adj_elem[1]

                if weight <= distance_left and (adj_vertex not in visited or visited[adj_vertex] < distance_left - weight):
                    visited[adj_vertex] = distance_left - weight
                    total_count(adj_vertex, distance_left - weight, visited)
            
            return len(visited)

        minDist = 1000000
        id = -1
        for v in range(n):
            print(v)
            visited = {}
            visited[v] = 10000
            
            n = total_count(v, distanceThreshold, visited)
            if n <= minDist:
                minDist = n
                id = v

        # return curr_min
        return id
            

File: LeetCode/1456-find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance/find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance.java
class Solution {
    HashMap <Integer, List<int []>> adj;
    public int findTheCity(int n, int[][] edges, int distanceThreshold) {
        int minNeighbors = Integer.MAX_VALUE, cityID = -1;
        adj = new HashMap <> ();

        //populate adj
        for (int i = 0; i < n; i++) adj.put(i, new ArrayList<int[]> ());

        for (int [] edge : edges) {
            adj.get(edge[0]).add(new int [] {edge[1], edge[2]});
            adj.get(edge[1]).add(new int [] {edge[0], edge[2]});
        } //for

        for (int i = 0; i < n; i++) {
            int [] seen = new int [n];
            for (int j = 0; j < n; j++) seen[j] = Integer.MIN_VALUE;
            seen[i] = Integer.MAX_VALUE;
            int neighbors = countNeighbors(i, distanceThreshold, seen);
            if (neighbors <= minNeighbors) {
                minNeighbors = neighbors;
                cityID = i;
            } //if
        } //for

        return cityID;
    } //findTheCity

    private int countNeighbors(int curr, int distLeft, int [] seen) {
        for (int [] next : adj.get(curr)) {
            int loc = next[0], dist = next[1];
            if (seen[loc] < (distLeft - dist) && dist <= distLeft) {
                seen[loc] = distLeft - dist;
                countNeighbors(loc, distLeft - dist, seen);
            } //if
        } //for

        int sol = 0;
        for (int city : seen) if (city >= 0) ++sol;
        return sol;
    } //countNeighbors
} //Sol